### ISSUE TITLE
Preserve annotation for nested types

### ISSUE THREAD

**Issue by kszucs** [pydantic/pydantic#782]
# Bug

Please complete:
* OS: **macOS**
* Python version `import sys; print(sys.version)`: **3.7.3**
* Pydantic version Master: https://github.com/samuelcolvin/pydantic/commit/5015a7e48bc869adf99b78eb38075951549e9ea7

**Please read the [docs](https://pydantic-docs.helpmanual.io/) and search through issues to
confirm your bug hasn't already been reported before.**

The following snippet fails with 
```
TypeError: The type of B.seq differs from the new default value; if you wish to change the type of this field, please use a type annotation
```

```python
import pydantic

from typing import List

from pydantic import BaseModel


class A(BaseModel):
    seq: List[int] = [0, 1, 2]


class B(A):
    seq = [4, 5]
```

@dmontagu I'd definitely expect it to pass. How about running `A.__fields__['seq'].validate()` on the new default value set at `B` instead of comparing the inferred and the previously annotated types?

This way it could ensure that the new default value is compatible with the previous type hints, without the need of advanced inferring.

**Comment by kszucs** [comment]
> @kszucs I think that is a good idea for an approach, but it might have the following problem: what if you convert an attribute originally annotated as a string to have a default value that is a float? I think that should probably be an error, but this change would cause it to be allowed (and it would be parsed to a string implicitly).
The current behaviour allows the following:

```python
class A(BaseModel):
    member: str = 3.5
        
assert type(A().member) is float
# but
assert type(A(member=3.5).member) is str
```

So I think the consistent behaviour would be to treat the default value as the input for the validation, thus allowing the coercion from float to string.  I suppose this would be the easiest too.
> 
> Another alternative would be to modify the behavior to check whether the elements of the list/dict/etc. actually match the annotation. It might involve special casing for `List`/`Dict`/etc., but even still might be preferable given how commonly used those types are.
Couldn't we reuse the validation system for this purpose?
> 
> A third alternative could be modifying the error message to just say that a type annotation must be added for the field (rather than the potentially confusing error that the inferred type was wrong). Given this would happen as soon as the class is created, I think this might be acceptable even if the exact logic is a little opaque.
Then modifying the default value without also repeating the annotation would not been possible for nested types. 
I'd prefer either the first or the second one. Please correct me if I'm wrong, but both seem to be implementable by the validation system, the only difference is that the first allows the coercion whereas the second does not. 
> 
> @kszucs @samuelcolvin Thoughts?

**Comment by samuelcolvin** [comment]
> Another alternative would be to modify the behavior to check whether the elements of the list/dict/etc. actually match the annotation.

This isn't currently possible with the validation system since it's designed to parse, not validate, so it "tries to coerce and errors if it can't" rather than "checking the types are correct".

The problem is that currently (unless you use `validate_all = True`) any default value is allowed - parsing/validation is not implemented on the default values.

Thus whatever we do here, it'll be slightly confusing compared to what we have at present.

I think the solution might be to just look at the outer type, currently lists without type annotations will cause the field to just require a list, eg:

```py
class Foo(BaseModel):
    x = [1, 2, 3]

print(Foo.__fields__)
#> {'x': <Field(x type=list default=[1, 2, 3])>}

Foo(x=['a', 'b'])  # ok

Foo(x=1)  # error: value is not a valid list 
```

In other words we don't do any clever checks to work out what type of list `[1, 2, 3]` - we just decide it's `List[Any]` since you wouldn't be bothered to add an annotation.

I would say we do roughly the same thing with inheritance: check the new default is a list/dict/set but not investigate it's contents.

**Comment by samuelcolvin** [comment]
Thus the following would be fine:

```py
from typing import List
from pydantic import BaseModel

class A(BaseModel):
    seq: List[int] = [0, 1, 2]

class B(A):
    seq = ['x', 'y']

Class B_Alternative(BaseModel):
    seq: List[int] = ['x', 'y']
```
`B.seq` would be of type `List[int]` with the (rather confusing, but acceptable default of) `['x', 'y']`, basically the same as `B_Alternative`.

**Comment by samuelcolvin** [comment]
Yes, I think the first error should remain, but not the second.

This should not be that difficult:
1. It involves only inspecting the type of the default, not any nested subtypes.
2. This is the same logic we already use to infer the type of a field without an annotation.

**Comment by samuelcolvin** [comment]
I'm sure you're not being blockheaded.

> Is the idea that if the annotated type is a container, we should only check for equality of the outermost type (e.g., the __origin__)?

Yes, exactly. But I haven't looked into this, I was just thinking out loud about a way to find a sensible compromise on this.

How about replacing `inferred.type_ != fields[var_name].type_` with something like:

```py
def annotation_matches_type(annotation_type: Type, value_type: Type):
    return annotation_type == value_type or getattr(annotation_type, '__origin__', None) == value_type
```
?

**Comment by samuelcolvin** [comment]
Okay, but I still think a function that fixes these cases and a few others wouldn't be that long or difficult to implement.

> For what it's worth, the @kszucs approach would probably handle all these cases neatly

Of calling `validate()`?

But wouldn't that confuse people much more? Eg. validators being called at load time and field defaults being validated in a few very specific circumstances but not otherwise.

I want to avoid confusing edge cases for those new to pydantic.

------------------

I've thought about this a lot and I don't see an elegant solution that works in all cases:

I think there are two options:
1. We require type annotations for all fields that override a field on a parent class
2. we implement a "best attempt" `annotation_matches_type` and modify the error message to more strongly encourage a type annotation; something like `it wasn't possible to confirm that the type of {value!r} matched {annotation_type}, please use an explicit type annotation on "{field_name}"`

I'd prefer option 2.

**Comment by samuelcolvin** [comment]
I'm happy with the current behaviour on main, not particularly inclined to change it, please open a new issue if you have a suggested change in v2.
